### DISCUSSION TITLE
USSD not working/crashing asterisk

### DISCUSSION THREAD

**Discussion by VasylSamoilov** [IchthysMaranatha/asterisk-chan-quectel#43] [Q&A] [answered]
I'm using Quectel EG25. Everything seems to work fine expect for USSD.
If I try to send USSD from asterisk CLI, it errors out 
`[quectel0] USSD queued for send`
`    -- [quectel0] Error sending USSD 0x7f5c001aa0`
`[Jan 16 21:02:40] ERROR[46643]: at_response.c:410 log_cmd_response_error: [quectel0] Error sending USSD 0x7f5c001aa0`

The reason for that that in plain-text USSD can't be requested in UCS2 encoding, which is the default for chan_quectel, only with GSM

`AT+CUSD=1,*111#,15`
`ERROR`
`AT+CSCS?`
`+CSCS: "UCS2"`

there are two options:
1) switch to GSM enconding
If I switch to `AT+CSCS="GSM"` asterisk-cli USSD command no longer errors out, but crashes asterisk
2) Send requests in UCS2 encoding (which is not possible from asterisk cli)
`AT+CUSD=1,"00230031003000300023"`, which again works fine, but crashes asterisk on reply

Output of async USSD commands is redirected to /dev/ttyUSB3 by default, which is modem port
This can be changed by telling quectel to switch output to AT port:
`AT+QURCCFG="urcport","usbat"`

If I do this, asterisk receive a response and crash on ussd reply due to`asterisk.service: Main process exited, code=killed, status=11/SEGV`
I would be happy to debug more to get USSD working.

**Comment by IchthysMaranatha** [comment]
I'm not able to check this as ussd commands are not working for me with minicom either with gsm encoding. Are they working for your provider in minicom? If yes, you could try commenting out this line 
https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/b3534eaedce52819ee3f61db6d191ed35177b955/at_command.c#L136
Recompile, restart modem and check.

**Reply by VasylSamoilov** [reply]
I am just using asterisk cli to do `quectel cmd...` to change settings on the fly. If I switch to GSM encoding it will accept USSD command but crash asterisk when getting reply. 
So, basically, there are two problems. First, for UCS2, USSD request in AT command should be encoded properly, which it doesn't do now.
Second, when you get USSD request to execute in any combination (UCS2 or GSM encoded), reply handling will crash the asterisk with SEGV.

**Reply by IchthysMaranatha** [reply]
First, something should work without Asterisk to correctly test, unless calls and sms work without Asterisk, they will not work with it either. Same with ussd. Unload Asterisk module and test with minicom first on same AT port. If it works, then one can proceed to fix it in Asterisk.

**Reply by IchthysMaranatha** [reply]
And have you tested with the change I suggested?

**Reply by VasylSamoilov** [reply]
Test with terminal:

+CPIN: READY
ATI
Quectel
EG25
Revision: EG25GGBR07A08M2G
AT+QURCCFG="urcport","usbat"
OK
AT+CSCS="GSM"
OK
AT+CUSD=1,*160#,15
OK

+CUSD: 0,"041A0438045704320441044C043A043804390020044704300441003A002000310034003A00300033003A00330030002000310037002F00300031002F0032003000320033",72
AT+CSCS="UCS2"
OK
AT+CUSD=1,*160#,15
ERROR
AT+CUSD=1,*160#
ERROR
AT+CUSD=1,"002A0031003600300023"
OK

+CUSD: 0,"041A0438045704320441044C043A043804390020044704300441003A002000310034003A00300035003A00330037002000310037002F00300031002F0032003000320033",72

**Reply by IchthysMaranatha** [reply]
Ok, it's working with your provider. Please make the change I suggested and build. Disconnect from terminal, restart modem, load changed module and check with Asterisk cli.

**Reply by VasylSamoilov** [reply]
I have made a change to  static const char cmd21[] = "AT+CSCS=\"GSM\"\r";
Otherwise it's hard to know that the setting value is - undefined behavior.
Now command works, but asterisk crashes on response processing.

`pi4*CLI> quectel ussd quectel0 *160#`
[quectel0] USSD queued for send
    -- [quectel0] Successfully sent USSD 0x7f88001b20
[Jan 17 15:38:28] NOTICE[1324]: at_response.c:347 at_response_ok: [quectel0] Successfully sent USSD 0x7f88001b20
pi4*CLI>
Disconnected from Asterisk server

I'll try to find out how to collect debug trace

**Comment by IchthysMaranatha** [comment]
May be a parsing issue I think. I can't check myself, as ussd does not work for me at all. Have you checked with another 2g/3g modem and chan_dongle? Do you get same response in terminal with 2g/3g modem?

**Reply by VasylSamoilov** [reply]
I don't have much of a variety of modems to test with, just quectel.
Here is debug log, I'm still thinking how to get gdb trace.

[Jan 17 16:29:16] DEBUG[1849]: at_response.c:202 at_response_ok: [quectel0] AT sent successfully
[Jan 17 16:29:16] DEBUG[1849]: at_queue.c:218 at_queue_remove_cmd: [quectel0] remove command 'AT' expected response 'OK' real 'OK' cmd 1/1 flags 0x03 from queue
[Jan 17 16:29:16] DEBUG[1849]: at_queue.c:67 at_queue_remove: [quectel0] remove task with 1 command(s) begin with 'AT' expected response 'OK' from queue
[Jan 17 16:29:18] DEBUG[1716]: res_pjsip_registrar.c:1377 check_expiration_thread: Woke up at 1673965758  Interval: 30
[Jan 17 16:29:18] DEBUG[1716]: res_pjsip_registrar.c:1384 check_expiration_thread: Expiring 0 contacts
[Jan 17 16:29:26] DEBUG[1849]: at_queue.c:124 at_queue_add: [quectel0] insert task with 1 commands begin with 'AT' expected response 'OK' after head of queue
[Jan 17 16:29:26] DEBUG[1849]: at_queue.c:245 at_queue_run: [quectel0] write command 'AT' expected response 'OK' length 3
[72B blob data]
[Jan 17 16:29:26] DEBUG[1849]: at_read.c:79 at_read: [quectel0] receive 6 byte, used 6, free 2042, read 0, write 6
[Jan 17 16:29:26] DEBUG[1849]: at_read.c:94 at_read: [quectel0] [
OK
]
[Jan 17 16:29:26] DEBUG[1849]: at_response.c:202 at_response_ok: [quectel0] AT sent successfully
[Jan 17 16:29:26] DEBUG[1849]: at_queue.c:218 at_queue_remove_cmd: [quectel0] remove command 'AT' expected response 'OK' real 'OK' cmd 1/1 flags 0x03 from queue
[Jan 17 16:29:26] DEBUG[1849]: at_queue.c:67 at_queue_remove: [quectel0] remove task with 1 command(s) begin with 'AT' expected response 'OK' from queue
[Jan 17 16:29:29] DEBUG[1832]: at_queue.c:124 at_queue_add: [quectel0] insert task with 1 commands begin with 'AT+CUSD' expected response 'OK' at tail of queue
[Jan 17 16:29:29] DEBUG[1832]: at_queue.c:245 at_queue_run: [quectel0] write command 'AT+CUSD' expected response 'OK' length 26
[95B blob data]
[Jan 17 16:29:29] DEBUG[1849]: at_read.c:79 at_read: [quectel0] receive 6 byte, used 6, free 2042, read 0, write 6
[Jan 17 16:29:29] DEBUG[1849]: at_read.c:94 at_read: [quectel0] [
OK
]
[Jan 17 16:29:29] NOTICE[1849]: at_response.c:347 at_response_ok: [quectel0] Successfully sent USSD 0x7f60074e30
[Jan 17 16:29:29] DEBUG[1849]: at_queue.c:218 at_queue_remove_cmd: [quectel0] remove command 'AT+CUSD' expected response 'OK' real 'OK' cmd 1/1 flags 0x00 from queue
[Jan 17 16:29:29] DEBUG[1849]: at_queue.c:67 at_queue_remove: [quectel0] remove task with 1 command(s) begin with 'AT+CUSD' expected response 'OK' from queue
[Jan 17 16:29:30] DEBUG[1849]: at_read.c:79 at_read: [quectel0] receive 31 byte, used 31, free 2017, read 0, write 31
[Jan 17 16:29:30] DEBUG[1849]: at_read.c:94 at_read: [quectel0] [
+CREG: 1,"6A20","1FC3967",2
]
[Jan 17 16:29:30] DEBUG[1849]: at_queue.c:124 at_queue_add: [quectel0] insert task with 1 commands begin with 'AT+COPS?' expected response 'OK' at tail of queue
[Jan 17 16:29:30] DEBUG[1849]: at_queue.c:245 at_queue_run: [quectel0] write command 'AT+COPS?' expected response 'OK' length 9
[78B blob data]
[Jan 17 16:29:30] DEBUG[1849]: at_read.c:79 at_read: [quectel0] receive 45 byte, used 45, free 2003, read 0, write 45
[Jan 17 16:29:30] DEBUG[1849]: at_read.c:94 at_read: [quectel0] [
+COPS: 0,0,"UA-KYIVSTAR KYIVSTAR",2

OK
]
[Jan 17 16:29:30] DEBUG[1849]: at_response.c:351 at_response_ok: [quectel0] Provider query successfully
[Jan 17 16:29:30] DEBUG[1849]: at_queue.c:218 at_queue_remove_cmd: [quectel0] remove command 'AT+COPS?' expected response 'OK' real 'OK' cmd 1/1 flags 0x01 from queue
[Jan 17 16:29:30] DEBUG[1849]: at_queue.c:67 at_queue_remove: [quectel0] remove task with 1 command(s) begin with 'AT+COPS?' expected response 'OK' from queue
[Jan 17 16:29:31] DEBUG[1849]: at_read.c:79 at_read: [quectel0] receive 9 byte, used 9, free 2039, read 0, write 9
asterisk.service: Main process exited, code=killed, status=11/SEGV
asterisk.service: Failed with result 'signal'.
asterisk.service: Consumed 18.220s CPU time.
asterisk.service: Scheduled restart job, restart counter is at 1.

**Reply by IchthysMaranatha** [reply]
One would have to place debug or log statements all through the ussd sending, receiving and parsing segments of code to identify exactly where in the code block it crashes. My suspicion is that it might be in the parsing and processing of received ussd.

**Reply by VasylSamoilov** [reply]
Sure. It was crashing near https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/b3534eaedce52819ee3f61db6d191ed35177b955/at_read.c#L82

after adding some debugging and sleep(1) here
https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/b3534eaedce52819ee3f61db6d191ed35177b955/ringbuffer.c#L52

it stopped crashing and let me debug more.

good news - ussd recieve works and encoding for at command payload should be UCS2, but encoding here is wrong:
https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/b3534eaedce52819ee3f61db6d191ed35177b955/at_command.c#L333

it doesn't match plain text as for GSM USSD, and doesn't match UCS2 hexified. So it actually never produces correct CUSD request.
I am not good at C to fix it.

When I send user command `asterisk -vvvvvvvrx 'quectel cmd quectel0 AT+CUSD=1,"002A0031003600300023",15'` it executes and decodes ussd `responce` correctly. Great success.

So, two problems

1. Wrong encoding of ussd code in `at_enqueue_ussd`
2. Crash in `rb_read_all_iov` unless there is a sleep in it

**Reply by VasylSamoilov** [reply]
Correction: Crash occurs due to `ERROR` response, nothing to do with sleep(). Somehow GSM7 encoded USSD code produces a OK reply from quectel that crashes afterwards in rb_read_all_iov on subsequent error. When supplied with proper UCS2 USSD code for AT+CUSD command it works just fine. Current workaround to invoke USSD from cli will be something like `asterisk -vrx 'quectel cmd quectel0 AT+CUSD=1,"002A0031003600300023",15'`

**Reply by IchthysMaranatha** [reply]
Good, you have done well. I'll look into the code and see if I find a fix for ussd ucs2 encoding with His help.

**Comment by IchthysMaranatha** [comment, marked as answer]
In your example, are you not passing diff commands with gsm and ucs2? With ucs2, you seem to passing # 100 # 
Anyway, I've added a fix using gsm, it changes encoding to gsm, passes ussd and then reverts to ucs2. Please check with this branch and let me know
https://github.com/IchthysMaranatha/asterisk-chan-quectel/tree/USSDTest
I've added a log here to see the command that is passed on to the modem, you can comment it out if the code works
https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/95b384e62f178a016f0a00651eb3260659813508/at_command.c#L379

**Reply by VasylSamoilov** [reply]
It will not work. GSM enconding for command should not be used here, it should be UCS2 hexified. 
`002A0031003600300023` is `*160#`

**Reply by IchthysMaranatha** [reply]
Does at+cscs="gsm" and at+cusd=1,"*111#",15 work for you on terminal? This does the same thing.

Anyway, as ussd does not work for me, not going to spend time on this. You will have to work it out yourself, or wait for someone else to help you.

**Reply by IchthysMaranatha** [reply]
Look at your opening post where you've used
AT+CUSD=1,*111#,15
AT+CUSD=1,"00230031003000300023"

**Reply by VasylSamoilov** [reply]
I stand corrected, it works.
[Jan 19 14:27:30] NOTICE[3806]: at_command.c:379 at_enqueue_ussd: USSD command passed is AT+CSCS="GSM";+CUSD=1,"*160#",15;+CSCS="UCS2"
[Jan 19 14:27:30] DEBUG[3806]: at_queue.c:124 at_queue_add: [quectel0] insert task with 1 commands begin with 'AT+CUSD' expected response 'OK' at tail of queue
[Jan 19 14:27:30] DEBUG[3806]: at_queue.c:245 at_queue_run: [quectel0] write command 'AT+CUSD' expected response 'OK' length 46
]Jan 19 14:27:30] DEBUG[3806]: at_queue.c:191 at_write: [quectel0] [AT+CSCS="GSM";+CUSD=1,"*160#",15;+CSCS="UCS2"
[Jan 19 14:27:30] DEBUG[3807]: at_read.c:79 at_read: [quectel0] receive 6 byte, used 6, free 2042, read 0, write 6
[Jan 19 14:27:30] DEBUG[3807]: at_read.c:94 at_read: [quectel0] [
OK
]
    -- [quectel0] Successfully sent USSD 0x7f60001a20

[Jan 19 14:27:33] DEBUG[3807]: at_read.c:79 at_read: [quectel0] receive 154 byte, used 154, free 1894, read 0, write 154
[Jan 19 14:27:33] DEBUG[3807]: at_read.c:94 at_read: [quectel0] [
+CUSD: 0,"041A0438045704320441044C043A043804390020044704300441003A002000310034003A00320037003A00330032002000310039002F00300031002F0032003000320033",72
]
 [quectel0] USSD DCS=2 (0: gsm7, 1: ascii, 2: ucs2)
 [quectel0] Got USSD type 0 'USSD Notify': 'Київський час: 14:27:32 19/01/2023'
[Jan 19 14:27:33] DEBUG[3807]: stasis.c:580 stasis_topic_create_with_detail: Creating topic. name: channel:1674131253.0, detail:

**Reply by VasylSamoilov** [reply]
Example that that you are referencing is my mistake due to incorrect copy/paste, I'm sorry. 
Correct example would be `002A0031003600300023` for UCS2 and `*160#` GSM mode.
In order to maintain coherency it would be good to encode USSD number in UCS2, in that case encoding switch will not be required. 
But it works anyway. Thank you.

**Reply by IchthysMaranatha** [reply]
You're welcome

**Reply by VasylSamoilov** [reply]
This will cover 90% of USSD use cases. 
From standards standpoint USSD implementation is not quite correct. This is just a note for anyone may be interested in re-implementing it in the future.
First of all, USSD should block all other operations on device from the time request is received until reply or error is reported back. There cannot be two ongoing USSDs in parallel.
Next, there is such thing as interactive USSD which implies user inputs after reply where a number or string should be entered to make a selection, effectively they are USSD menus (which are pretty much an equivalent to voice menus), but they are rare.

**Comment by RoEdAl** [comment]
My [fork](https://github.com/RoEdAl/asterisk-chan-quectel) of `chan_quectel` now sends an receives `UCS-2`-encoded USSD strings without additional `AT+CSCS` commands.

**Comment by hello-22-world** [comment]
I'm still encountering the same USSD issue in the new release. The error message "Error sending USSD 0x74e268004820" appears, and then Asterisk hangs, preventing calls and functionality.

Could you please provide guidance on how to resolve this issue?

[quectel0] USSD queued for send
    -- [quectel0] Error sending USSD 0x74e268004820
[Oct 20 00:06:43] ERROR[3079]: at_response.c:410 log_cmd_response_error: [quectel0] Error sending USSD 0x74e268004820

**Comment by IchthysMaranatha** [comment]
Are you using code from this branch? https://github.com/IchthysMaranatha/asterisk-chan-quectel/tree/USSDTest

**Reply by hello-22-world** [reply]
I'm using the code from the main branch, but this tree doesn't seem to be up-to-date?

**Reply by IchthysMaranatha** [reply]
No, changes have not been made to main brn

**Reply by hello-22-world** [reply]
Is it possible to merge these tree into the main branch? Thank you so much for your work.

**Reply by IchthysMaranatha** [reply]
No changes in main brn after this brn except makefile for openwrt, you can use the generator repo if using with openwrt. USSD changes are minor https://github.com/IchthysMaranatha/asterisk-chan-quectel/commit/e60443c79b56d3236a2af1e3267fa9fa80916665 and https://github.com/IchthysMaranatha/asterisk-chan-quectel/commit/10cec0ba8016e23d6670f8ea6657eb4a178f399c

**Reply by hello-22-world** [reply]
The USSD data/string returned by the carrier is causing Asterisk to hang and restart.
(The quectel2 dongle experiences a hang problem, while other dongles produce a garbled output of characters)
When it doesn't hang, I'm only seeing a jumbled mess of code with lots of "¥Δå@Ψ" symbols. 
Is there any way to resolve this issue? 

it is not work to decode the BASE64_DECODE code:
exten => ussd,1,Verbose(Incoming USSD: ${BASE64_DECODE(${USSD_BASE64})})
exten => ussd,n,System(echo '${STRFTIME(${EPOCH},,%Y-%m-%d %H:%M:%S)} - ${QUECTELNAME}: ${BASE64_DECODE(${USSD_BASE64})}' >> /var/log/asterisk/ussd.txt)
exten => ussd,n,Hangup()

Asterisk*CLI> quectel cmd quectel2 AT+CUSD=1,"00230031003000300023",15
Disconnected from Asterisk server
Asterisk cleanly ending (0).
Executing last minute cleanups

**Reply by IchthysMaranatha** [reply]
Again, are you using ussd branch or main branch?

**Reply by hello-22-world** [reply]
> Again, are you using ussd branch or main branch?

i am using the ussd branch now

**Reply by IchthysMaranatha** [reply]
What do logs show?

**Reply by hello-22-world** [reply]
> What do logs show?

 -- Executing [ussd@incoming-mobile:1] Verbose("Local/ussd@incoming-mobile-00000002;1", "Incoming USSD: @(£¡ù Σ@n@,.....


and it will hang randomly on one of the dongle : 

Asterisk*CLI> quectel ussd quectel2 *111#
[quectel2] USSD queued for send
[Oct 27 11:11:34] NOTICE[6172]: at_command.c:379 at_enqueue_ussd: USSD command passed is AT+CSCS="GSM";+CUSD=1,"*111#",15;+CSCS="UCS2"
    -- [quectel2] Successfully sent USSD 0x7988b0001c00
[Oct 27 11:11:34] NOTICE[6168]: at_response.c:347 at_response_ok: [quectel2] Successfully sent USSD 0x7988b0001c00
Asterisk*CLI>
Disconnected from Asterisk server
Asterisk cleanly ending (0).
Executing last minute cleanups

**Comment by IchthysMaranatha** [comment]
As my operator does not support ussd, no way for me to test USSD myself

Unload asterisk, use minicom to connect to modem and then run 
AT+CSCS="GSM";+CUSD=1,"111#",15;+CSCS="UCS2"
What do you get in response?

**Reply by hello-22-world** [reply]
> As my operator does not support ussd, no way for me to test USSD myself
> 
> Unload asterisk, use minicom to connect to modem and then run AT+CSCS="GSM";+CUSD=1,"111#",15;+CSCS="UCS2" What do you get in response?

I'm able to get responses sometimes, but they appear as scrambled text or symbols.

-- Executing [ussd@incoming-mobile:1] Verbose("Local/ussd@incoming-mobile-00000002;1", "Incoming USSD: @(£¡ù Σ@n@,.....

**Reply by IchthysMaranatha** [reply]
No, not with asterisk, unload module from asterisk and try with minicom connected to modem command port

**Reply by hello-22-world** [reply]
The text is in UCS-2 format, but it has a leading zero (0) that needs to be removed before decoding. I tried decoding it with ChatGPT, and it was able to successfully decode it as follows:

it is return : "+CUSD: 0,"005400680061006E006B007300200066006F00720020007500730069006E00670020........

+CUSD: 0, This is a standard prefix for USSD responses.
0054: This is the Unicode code point for the letter "T".
0068: This is the Unicode code point for the letter "h".
0061: This is the Unicode code point for the letter "a".
006E: This is the Unicode code point for the letter "n".
006B: This is the Unicode code point for the letter "k".
0073: This is the Unicode code point for the letter "s".
0020: This is the Unicode code point for a space.
0066: This is the Unicode code point for the letter "f".
006F: This is the Unicode code point for the letter "o".
0072: This is the Unicode code point for the letter "r".
0020: This is the Unicode code point for a space.
0075: This is the Unicode code point for the letter "u".
0073: This is the Unicode code point for the letter "s".
0069: This is the Unicode code point for the letter "i".
006E: This is the Unicode code point for the letter "n".
0067: This is the Unicode code point for the letter "g".
0020: This is the Unicode code point for a space.

**Reply by IchthysMaranatha** [reply]
Ok good, try this, insert foll block of code after line https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/280257455edf7a28a21d327643161fab7294b729/at_response.c#L1704
```
    char result [1000]  ;
    int len= strlen(cusd);
    for (int i=2;i<len;i+=4){
        result[i/2-1]= cusd[i];
        result [i/2]= cusd[i+1];
        if(i==len-2) result [i/2+1]= '\0';
   
    }
    cusd=result;
```

**Reply by hello-22-world** [reply]
> Ok good, try this, insert foll block of code after line
> 
> https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/280257455edf7a28a21d327643161fab7294b729/at_response.c#L1704
> 
> ```
>     char result [1000]  ;
>     int len= strlen(cusd);
>     for (int i=2;i<len;i+=4){
>         result[i/2-1]= cusd[i];
>         result [i/2]= cusd[i+1];
>         if(i==len-2) result [i/2+1]= '\0';
>    
>     }
>     cusd=result;
> ```

Thank you so much for your help, but I'm getting the following error when I try to compile:

at_response.c: In function ‘at_response_cusd’:
at_response.c:1706:13: error: ‘len’ redeclared as different kind of symbol
 1706 |         int len= strlen(cusd);

I've modified the code as follows, but I'm still unable to resolve the issue:

 ast_verb(1, "[%s] USSD DCS=%d (0: gsm7, 1: ascii, 2: ucs2)\n", PVT_ID(pvt), dcs);

    char result[1000];
    size_t str_length = strlen(cusd);  // Changed from 'len' to 'str_length'
    for (int i = 2; i < str_length; i += 4) {
        result[i/2-1] = cusd[i];
        result[i/2] = cusd[i+1];
        if (i == str_length-2) result[i/2+1] = '\0';
    }
    cusd = result;

**Reply by hello-22-world** [reply]
I'm not sure if there's an issue with the code, as I'm not very experienced with coding. 
However, the AI seems to have fixed the problem by suggesting the following code:

```c
static int at_response_cusd(struct pvt* pvt, char* str, size_t len)
{
    static const char* const types[] = {
        "USSD Notify",
        "USSD Request",
        "USSD Terminated by network",
        "Other local client has responded",
        "Operation not supported",
        "Network time out",
    };
    ssize_t     res;
    int         type;
    char*       cusd;
    int         dcs;
    char        cusd_utf8_str[1024];
    char        text_base64[16384];
    char        typebuf[2];
    const char* typestr;
    
    manager_event_message("QuectelNewCUSD", PVT_ID(pvt), str);
    
    if (at_parse_cusd(str, &type, &cusd, &dcs))
    {
        ast_verb(1, "[%s] Error parsing CUSD: '%.*s'\n", PVT_ID(pvt), (int)len, str);
        return -1;
    }

    if (type < 0 || type >= (int)ITEMS_OF(types))
    {
        ast_log(LOG_WARNING, "[%s] Unknown USSD type: %d\n", PVT_ID(pvt), type);
    }

    typestr = enum2str(type, types, ITEMS_OF(types));
    typebuf[0] = type + '0';
    typebuf[1] = 0;

    // Store original DCS for logging
    int original_dcs = dcs;
    
    // Properly handle DCS according to 3GPP TS 23.038
    int coding_scheme = 0;
    if (dcs & 0x40) {  // Check if bit 6 is set
        // Bits 3,2 contain the coding scheme
        coding_scheme = (dcs & 0x0C) >> 2;
        if (coding_scheme == 3) {
            coding_scheme = 0;  // Default to GSM 7-bit if reserved value
        }
    } else if (dcs & 0x20) {  // Bit 5 set - Message waiting indication group
        coding_scheme = 0;     // GSM 7-bit default
    } else if ((dcs & 0xF0) == 0) {  // General Data Coding group
        coding_scheme = dcs & 0x0C;   // Bits 3,2 contain coding scheme
        coding_scheme >>= 2;
    } else {
        coding_scheme = 0;     // Default to GSM 7-bit for unhandled cases
    }

    ast_verb(1, "[%s] USSD Original DCS=0x%02x, Decoded coding scheme=%d (0: gsm7, 1: ascii, 2: ucs2)\n", 
             PVT_ID(pvt), original_dcs, coding_scheme);

    // Remove quotes if present
    if (cusd[0] == '"') {
        cusd++;  // Skip opening quote
        size_t cusd_len = strlen(cusd);
        if (cusd_len > 0 && cusd[cusd_len - 1] == '"') {
            cusd[cusd_len - 1] = '\0';  // Remove closing quote
        }
    }

    // Check if the string appears to be UCS-2 encoded (starts with 00xx00xx pattern)
    int appears_ucs2 = 0;
    if (strlen(cusd) >= 8) {
        if (strncmp(cusd, "00", 2) == 0 && 
            strncmp(cusd + 4, "00", 2) == 0) {
            appears_ucs2 = 1;
            ast_verb(1, "[%s] Content appears to be UCS-2 encoded\n", PVT_ID(pvt));
        }
    }

    // Handle as UCS-2 if either DCS indicates it or the content appears to be UCS-2
    if (coding_scheme == 2 || appears_ucs2) {
        // Convert hex string to binary UCS-2
        char binary_data[2048];
        size_t binary_len = 0;
        
        // Process hex string two characters at a time
        for (size_t i = 0; i < strlen(cusd); i += 2) {
            if (i + 1 < strlen(cusd)) {
                char hex[3] = {cusd[i], cusd[i + 1], '\0'};
                int value;
                if (sscanf(hex, "%x", &value) == 1) {
                    binary_data[binary_len++] = (char)value;
                }
            }
        }
        
        // Convert UCS-2 to UTF-8
        res = ucs2_to_utf8((uint16_t*)binary_data, binary_len / 2, 
                          cusd_utf8_str, sizeof(cusd_utf8_str) - 1);
    }
    else if (coding_scheme == 0 && !appears_ucs2) { // GSM-7
        uint16_t out_ucs2[1024];
        int cusd_nibbles = unhex(cusd, cusd);
        res = gsm7_unpack_decode(cusd, cusd_nibbles, out_ucs2, 
                                sizeof(out_ucs2) / 2, 0, 0, 0);
        if (res < 0) {
            return -1;
        }
        res = ucs2_to_utf8(out_ucs2, res, cusd_utf8_str, 
                          sizeof(cusd_utf8_str) - 1);
    }
    else if (coding_scheme == 1) { // ASCII (8-bit)
        res = strlen(cusd);
        if (res > sizeof(cusd_utf8_str) - 1) {
            res = -1;
        }
        else {
            memcpy(cusd_utf8_str, cusd, res);
        }
    }
    else {
        res = -1;
    }

    if (res < 0) {
        ast_log(LOG_WARNING, "[%s] Failed to decode USSD message (DCS=0x%02x)\n", 
                PVT_ID(pvt), original_dcs);
        return -1;
    }

    cusd_utf8_str[res] = '\0';
    ast_verb(1, "[%s] Got USSD type %d '%s': '%s'\n", PVT_ID(pvt), type, typestr, cusd_utf8_str);

    ast_base64encode(text_base64, (unsigned char*)cusd_utf8_str, res, sizeof(text_base64));

    manager_event_new_ussd(PVT_ID(pvt), cusd_utf8_str);
    manager_event_message("QuectelNewUSSDBase64", PVT_ID(pvt), text_base64);

    {
        channel_var_t vars[] =
        {
            { "USSD_TYPE", typebuf },
            { "USSD_TYPE_STR", ast_strdupa(typestr) },
            { "USSD", cusd_utf8_str },
            { "USSD_BASE64", text_base64 },
            { NULL, NULL },
        };
        start_local_channel(pvt, "ussd", "ussd", vars);
    }
    return 0;
}
``` 


 [quectel0] USSD Original DCS=0x00, Decoded coding scheme=0 (0: gsm7, 1: ascii, 2: ucs2)
 [quectel0] Content appears to be UCS-2 encoded
 [quectel0] Got USSD type 0 'USSD Notify': 'Thanks for using xxxxx service, a short message will be sent to you shortly.'

**Reply by IchthysMaranatha** [reply]
Please use something else like len1 or len2 in all the 3 places len appears in the block but strlen should remain as it is

**Reply by IchthysMaranatha** [reply]
Sorry, did not see later reply, wait

**Reply by IchthysMaranatha** [reply]
Ok good, so you have resolved the problem?

**Reply by hello-22-world** [reply]
> Ok good, so you have resolved the problem?

The problem appears to be fixed. Could you please review the AI modifications made to the code above to see if any issues remain?

**Reply by IchthysMaranatha** [reply]
I'm not really a developer myself, just tinker around to get things working. Anyway, this portion of code completely restricted to uusd alone so it would not cause any other issues and if your ussd is working fine, then great! Well done.

**Reply by hello-22-world** [reply]
I apologize for the issue. Asterisk is still crashing after sending the USSD code, and the kernel is reporting the following error:

kernel: asterisk[9145]: segfault at 8 ip 0000764ca4aa1892 sp 0000764c9d87aaa0 error 4 in chan_quectel.so[764ca4a9b000+26000] likely on CPU 0 (core 0, socket 0)

Could you please advise on how to debug this issue?

**Reply by IchthysMaranatha** [reply]
Ok, first try this, switch back to original ussdtest branch code without any modifications and insert these lines after 
https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/280257455edf7a28a21d327643161fab7294b729/at_response.c#L1704
``` 
ast_log (LOG_ERROR, "dcs obtained is %d\n", dcs);
ast_log (LOG_ERROR, "cusd string obtained is %s\n", cusd);
dcs=2;
 ```
If it crashes, show me the log and we'll try and find the cause with further changes in code

**Reply by hello-22-world** [reply]
> Ok, first try this, switch back to original ussdtest branch code without any modifications and insert these lines after
> 
> https://github.com/IchthysMaranatha/asterisk-chan-quectel/blob/280257455edf7a28a21d327643161fab7294b729/at_response.c#L1704
> 
> ```
> ast_log (LOG_ERROR, "dcs obtained is %d\n", dcs);
> ast_log (LOG_ERROR, "cusd string obtained is %s\n", cusd);
> dcs=2;
> ```
> 
> If it crashes, show me the log and we'll try and find the cause with further changes in code

While monitoring, the issue appears to be caused by server or network problems (high server load or weak signal).  
After the message "at_response.c:2159 at_response: [quectel0] Got Response for user's command: 'OK'".
The Asterisk system crashes if no response is received (About 30s) . 
How can I increase the response timeout, or allow the process to continue even without a response? 
Thank you for your assistance.

**Reply by IchthysMaranatha** [reply]
Ah then, you should be resolving those issues rather than tinkering with this code. I haven't looked at it deeply but change required as you desire may not even be in this code but in one of the C libraries used.

**Comment by IchthysMaranatha** [comment]
Also, have you tried this fork created by someone else
https://github.com/RoEdAl/asterisk-chan-quectel

**Reply by hello-22-world** [reply]
> Also, have you tried this fork created by someone else https://github.com/RoEdAl/asterisk-chan-quectel

I tested this fork today and was able to get the correct USSD response, but I'm encountering an error when trying to receive SMS messages.

**Comment by Roseclear** [comment]
Some provider response ussd encode method is gsm, but actually, it use ucs2 encoding.
That make the decoder use gsm decode, and output a bunch of garbage

I want to suggest some way to patch that here:
1. Remove gsm encode/decode, make everything use ucs2 as default, auto translate the ussd code to ucs2 before send
2. Or, keep the gsm encode/decode, but instead blindly trust the response header about used encoder, check the actual content to know what is the format of response
